opt_einsum/paths.py

```diff
@@ -272,3 +279,210 @@ def greedy(input_sets, output_set, idx_dict, memory_limit):
         path_cost += best[0][1]
 
     return path
+
+
+def ssa_to_linear(ssa_path):
+    """
+    Convert a path with static single assignment ids to a path with recycled
+    linear ids. For example::
+
+        >>> ssa_to_linear([(0, 3), (2, 4), (1, 5)])
+        [(0, 3), (1, 2), (0, 1)]
+    """
+    ids = np.arange(1 + max(map(max, ssa_path)), dtype=np.int32)
+    path = []
+    for ssa_ids in ssa_path:
+        path.append(tuple(int(ids[ssa_id]) for ssa_id in ssa_ids))
+        for ssa_id in ssa_ids:
+            ids[ssa_id:] -= 1
+    return path
+
+
+def linear_to_ssa(path):
+    """
+    Convert a path with recycled linear ids to a path with static single
+    assignment ids. For example::
+
+        >>> linear_to_ssa([(0, 3), (1, 2), (0, 1)])
+        [(0, 3), (2, 4), (1, 5)]
+    """
+    num_inputs = sum(map(len, path)) - len(path) + 1
+    linear_to_ssa = list(range(num_inputs))
+    new_ids = itertools.count(num_inputs)
+    ssa_path = []
+    for ids in path:
+        ssa_path.append(tuple(linear_to_ssa[id_] for id_ in ids))
+        for id_ in sorted(ids, reverse=True):
+            del linear_to_ssa[id_]
+        linear_to_ssa.append(next(new_ids))
+    return ssa_path
+
+
+def _get_candidate(output, sizes, remaining, costs, dim_ref_counts, k1, k2):
+    either = k1 | k2
+    two = k1 & k2
+    one = either - two
+    k12 = (either & output) | (two & dim_ref_counts[3]) | (one & dim_ref_counts[2])
+    cost = costs[k1] + costs[k2] + helpers.compute_size_by_dict(k12, sizes)
```

Comment:
I think I would prefer two things here:

1. ``size12 - size1 - size2`` I still think is the best default, it works best for all the tricky contractions I have (current version with ``(size1 -size2) / 2`` varies between 1-4x worse), and is quite natural since it is just proportional to the amount of memory removed.

2. Maybe this was done for efficiency, but I think its a bit confusing having the cost logic spread out over several places, i.e. ``costs[k1]`` and ``costs[k1]`` are halved and negative here, but that's not very clear. If the line was explicitly ``cost = footprint[k12] - footprint[1] - footprint[2]`` I think that would be beneficial, particularly if it comes to implementing a custom cost function.

Reply:
Done, thanks for reviewing. I believe this should now be very close to the `greedy` implementation.